// kepler.gl/dist/components/side-panel/layer-panel/vis-config-slider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.VisConfigSlider = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = require("../../common/styled-components");

var _rangeSlider = _interopRequireDefault(require("../../common/range-slider"));

var _reactIntl = require("react-intl");

// Copyright (c) 2020 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
var propTypes = {
  layer: _propTypes["default"].object.isRequired,
  property: _propTypes["default"].string.isRequired,
  onChange: _propTypes["default"].func.isRequired,
  label: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].bool, _propTypes["default"].func]),
  range: _propTypes["default"].arrayOf(_propTypes["default"].number).isRequired,
  step: _propTypes["default"].number,
  isRanged: _propTypes["default"].bool,
  disabled: _propTypes["default"].bool,
  inputTheme: _propTypes["default"].bool
};

var VisConfigSlider = function VisConfigSlider(_ref) {
  var config = _ref.layer.config,
      property = _ref.property,
      label = _ref.label,
      range = _ref.range,
      step = _ref.step,
      isRanged = _ref.isRanged,
      disabled = _ref.disabled,
      _onChange2 = _ref.onChange,
      inputTheme = _ref.inputTheme;
  return _react["default"].createElement(_styledComponents.SidePanelSection, {
    disabled: Boolean(disabled)
  }, label ? _react["default"].createElement(_styledComponents.PanelLabel, null, typeof label === 'string' ? _react["default"].createElement(_reactIntl.FormattedMessage, {
    id: label
  }) : typeof label === 'function' ? _react["default"].createElement(_reactIntl.FormattedMessage, {
    id: label(config)
  }) : _react["default"].createElement(_reactIntl.FormattedMessage, {
    id: "property.".concat(property)
  })) : null, _react["default"].createElement(_rangeSlider["default"], {
    range: range,
    value0: isRanged ? config.visConfig[property][0] : range[0],
    value1: isRanged ? config.visConfig[property][1] : config.visConfig[property],
    step: step,
    isRanged: Boolean(isRanged),
    onChange: function onChange(value) {
      return _onChange2((0, _defineProperty2["default"])({}, property, isRanged ? value : value[1]));
    },
    inputTheme: inputTheme,
    showInput: true
  }));
};

exports.VisConfigSlider = VisConfigSlider;
VisConfigSlider.propTypes = propTypes;
var _default = VisConfigSlider;
exports["default"] = _default;
