// kepler.gl/dist/components/side-panel/common/dataset-info.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = DatasetInfoFactory;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _d3Format = require("d3-format");

var _reactIntl = require("react-intl");

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 11px;\n  color: ", ";\n  padding-left: 19px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var numFormat = (0, _d3Format.format)(',');

var StyledDataRowCount = _styledComponents["default"].div(_templateObject(), function (props) {
  return props.theme.subtextColor;
});

function DatasetInfoFactory() {
  var DatasetInfo = function DatasetInfo(_ref) {
    var dataset = _ref.dataset;
    return _react["default"].createElement(StyledDataRowCount, {
      className: "source-data-rows"
    }, _react["default"].createElement(_reactIntl.FormattedMessage, {
      id: 'datasetInfo.rowCount',
      values: {
        rowCount: numFormat(dataset.allData.length)
      }
    }));
  };

  return DatasetInfo;
}
