// kepler.gl/dist/constants/user-guides.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GITHUB_USER_GUIDE = exports.GITHUB_BUG_REPORT = exports.GUIDES_FILE_FORMAT = exports.GITHUB_EXPORT_HTML_MAP_MODES = exports.MAPBOX_ACCESS_TOKEN = exports.GITHUB_ADD_DATA_TO_MAP = exports.GITHUB_EXPORT_HTML_MAP = void 0;
// Copyright (c) 2020 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
var GITHUB_EXPORT_HTML_MAP = 'https://github.com/keplergl/kepler.gl/blob/master/docs/user-guides/k-save-and-export.md#2save-and-export-current-map';
exports.GITHUB_EXPORT_HTML_MAP = GITHUB_EXPORT_HTML_MAP;
var GITHUB_ADD_DATA_TO_MAP = 'https://github.com/keplergl/kepler.gl/blob/master/docs/api-reference/actions/actions.md#adddatatomap';
exports.GITHUB_ADD_DATA_TO_MAP = GITHUB_ADD_DATA_TO_MAP;
var MAPBOX_ACCESS_TOKEN = 'https://docs.mapbox.com/help/how-mapbox-works/access-tokens/'; // TODO: create new readme section

exports.MAPBOX_ACCESS_TOKEN = MAPBOX_ACCESS_TOKEN;
var GITHUB_EXPORT_HTML_MAP_MODES = 'https://github.com/keplergl/kepler.gl/blob/master/docs/user-guides/k-save-and-export.md#2save-and-export-current-map';
exports.GITHUB_EXPORT_HTML_MAP_MODES = GITHUB_EXPORT_HTML_MAP_MODES;
var GUIDES_FILE_FORMAT = 'https://github.com/keplergl/kepler.gl/blob/master/docs/user-guides/b-kepler-gl-workflow/a-add-data-to-the-map.md#supported-file-formats';
exports.GUIDES_FILE_FORMAT = GUIDES_FILE_FORMAT;
var GITHUB_BUG_REPORT = 'https://github.com/keplergl/kepler.gl/issues/new?assignees=&labels=bug&template=bug_report.md&title=%5BBug%5D';
exports.GITHUB_BUG_REPORT = GITHUB_BUG_REPORT;
var GITHUB_USER_GUIDE = 'https://github.com/keplergl/kepler.gl/blob/master/docs/user-guides/a-introduction.md';
exports.GITHUB_USER_GUIDE = GITHUB_USER_GUIDE;
