// kepler.gl/dist/components/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  TimeRangeSlider: true,
  TimeRangeSliderFactory: true,
  KeplerGl: true,
  injectComponents: true,
  KeplerGlFactory: true,
  SidePanelFactory: true,
  PanelTitleFactory: true,
  MapContainerFactory: true,
  BottomWidgetFactory: true,
  ModalContainerFactory: true,
  PlotContainerFactory: true,
  PanelHeaderFactory: true,
  SaveExportDropdownFactory: true,
  PanelHeaderDropdownFactory: true,
  PanelHeaderAction: true,
  CollapseButtonFactory: true,
  SidebarFactory: true,
  PanelToggleFactory: true,
  AddDataButtonFactory: true,
  LayerManagerFactory: true,
  LayerPanelFactory: true,
  LayerPanelHeaderFactory: true,
  LayerConfiguratorFactory: true,
  SourceDataCatalogFactory: true,
  SourceDataSelectorFactory: true,
  DatasetTitleFactory: true,
  DatasetInfoFactory: true,
  DatasetTagFactory: true,
  FilterManagerFactory: true,
  FilterPanelFactory: true,
  InteractionManagerFactory: true,
  BrushConfigFactory: true,
  TooltipConfigFactory: true,
  MapManagerFactory: true,
  LayerGroupSelectorFactory: true,
  MapStyleSelectorFactory: true,
  CustomPanelsFactory: true,
  MapPopoverFactory: true,
  MapControlFactory: true,
  LayerHoverInfoFactory: true,
  CoordinateInfoFactory: true,
  ModalDialogFactory: true,
  DeleteDatasetModalFactory: true,
  DataTableModalFactory: true,
  LoadDataModalFactory: true,
  ExportImageModalFactory: true,
  ExportDataModalFactory: true,
  AddMapStyleModalFactory: true,
  ExportMapModalFactory: true,
  ModalTabsFactory: true,
  LoadStorageMapFactory: true,
  AnimationControlFactory: true,
  SpeedControlFactory: true,
  AnimationPlaybacksFactory: true,
  FloatingTimeDisplayFactory: true,
  DatasetSquare: true,
  TimeWidgetFactory: true,
  SingleSelectFilterFactory: true,
  MultiSelectFilterFactory: true,
  TimeRangeFilterFactory: true,
  RangeFilterFactory: true,
  injector: true,
  withState: true,
  CloudTile: true,
  FileUploadFactory: true,
  FileUpload: true,
  DatasetLabel: true,
  ItemSelector: true,
  FieldSelector: true,
  RangeSlider: true,
  Modal: true,
  ModalFooter: true,
  ModalTitle: true,
  AppLogo: true,
  Switch: true,
  LoadingSpinner: true,
  LoadingDialog: true,
  Icons: true
};
Object.defineProperty(exports, "TimeRangeSliderFactory", {
  enumerable: true,
  get: function get() {
    return _timeRangeSlider["default"];
  }
});
Object.defineProperty(exports, "KeplerGl", {
  enumerable: true,
  get: function get() {
    return _container["default"];
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _container["default"];
  }
});
Object.defineProperty(exports, "injectComponents", {
  enumerable: true,
  get: function get() {
    return _container.injectComponents;
  }
});
Object.defineProperty(exports, "KeplerGlFactory", {
  enumerable: true,
  get: function get() {
    return _keplerGl["default"];
  }
});
Object.defineProperty(exports, "SidePanelFactory", {
  enumerable: true,
  get: function get() {
    return _sidePanel["default"];
  }
});
Object.defineProperty(exports, "PanelTitleFactory", {
  enumerable: true,
  get: function get() {
    return _sidePanel.PanelTitleFactory;
  }
});
Object.defineProperty(exports, "MapContainerFactory", {
  enumerable: true,
  get: function get() {
    return _mapContainer["default"];
  }
});
Object.defineProperty(exports, "BottomWidgetFactory", {
  enumerable: true,
  get: function get() {
    return _bottomWidget["default"];
  }
});
Object.defineProperty(exports, "ModalContainerFactory", {
  enumerable: true,
  get: function get() {
    return _modalContainer["default"];
  }
});
Object.defineProperty(exports, "PlotContainerFactory", {
  enumerable: true,
  get: function get() {
    return _plotContainer["default"];
  }
});
Object.defineProperty(exports, "PanelHeaderFactory", {
  enumerable: true,
  get: function get() {
    return _panelHeader["default"];
  }
});
Object.defineProperty(exports, "SaveExportDropdownFactory", {
  enumerable: true,
  get: function get() {
    return _panelHeader.SaveExportDropdownFactory;
  }
});
Object.defineProperty(exports, "PanelHeaderDropdownFactory", {
  enumerable: true,
  get: function get() {
    return _panelHeader.PanelHeaderDropdownFactory;
  }
});
Object.defineProperty(exports, "PanelHeaderAction", {
  enumerable: true,
  get: function get() {
    return _panelHeaderAction["default"];
  }
});
Object.defineProperty(exports, "CollapseButtonFactory", {
  enumerable: true,
  get: function get() {
    return _sideBar.CollapseButtonFactory;
  }
});
Object.defineProperty(exports, "SidebarFactory", {
  enumerable: true,
  get: function get() {
    return _sideBar["default"];
  }
});
Object.defineProperty(exports, "PanelToggleFactory", {
  enumerable: true,
  get: function get() {
    return _panelToggle["default"];
  }
});
Object.defineProperty(exports, "AddDataButtonFactory", {
  enumerable: true,
  get: function get() {
    return _layerManager.AddDataButtonFactory;
  }
});
Object.defineProperty(exports, "LayerManagerFactory", {
  enumerable: true,
  get: function get() {
    return _layerManager["default"];
  }
});
Object.defineProperty(exports, "LayerPanelFactory", {
  enumerable: true,
  get: function get() {
    return _layerPanel["default"];
  }
});
Object.defineProperty(exports, "LayerPanelHeaderFactory", {
  enumerable: true,
  get: function get() {
    return _layerPanelHeader["default"];
  }
});
Object.defineProperty(exports, "LayerConfiguratorFactory", {
  enumerable: true,
  get: function get() {
    return _layerConfigurator["default"];
  }
});
Object.defineProperty(exports, "SourceDataCatalogFactory", {
  enumerable: true,
  get: function get() {
    return _sourceDataCatalog["default"];
  }
});
Object.defineProperty(exports, "SourceDataSelectorFactory", {
  enumerable: true,
  get: function get() {
    return _sourceDataSelector["default"];
  }
});
Object.defineProperty(exports, "DatasetTitleFactory", {
  enumerable: true,
  get: function get() {
    return _datasetTitle["default"];
  }
});
Object.defineProperty(exports, "DatasetInfoFactory", {
  enumerable: true,
  get: function get() {
    return _datasetInfo["default"];
  }
});
Object.defineProperty(exports, "DatasetTagFactory", {
  enumerable: true,
  get: function get() {
    return _datasetTag["default"];
  }
});
Object.defineProperty(exports, "FilterManagerFactory", {
  enumerable: true,
  get: function get() {
    return _filterManager["default"];
  }
});
Object.defineProperty(exports, "FilterPanelFactory", {
  enumerable: true,
  get: function get() {
    return _filterPanel["default"];
  }
});
Object.defineProperty(exports, "InteractionManagerFactory", {
  enumerable: true,
  get: function get() {
    return _interactionManager["default"];
  }
});
Object.defineProperty(exports, "BrushConfigFactory", {
  enumerable: true,
  get: function get() {
    return _brushConfig["default"];
  }
});
Object.defineProperty(exports, "TooltipConfigFactory", {
  enumerable: true,
  get: function get() {
    return _tooltipConfig["default"];
  }
});
Object.defineProperty(exports, "MapManagerFactory", {
  enumerable: true,
  get: function get() {
    return _mapManager["default"];
  }
});
Object.defineProperty(exports, "LayerGroupSelectorFactory", {
  enumerable: true,
  get: function get() {
    return _mapLayerSelector["default"];
  }
});
Object.defineProperty(exports, "MapStyleSelectorFactory", {
  enumerable: true,
  get: function get() {
    return _mapStyleSelector["default"];
  }
});
Object.defineProperty(exports, "CustomPanelsFactory", {
  enumerable: true,
  get: function get() {
    return _customPanel["default"];
  }
});
Object.defineProperty(exports, "MapPopoverFactory", {
  enumerable: true,
  get: function get() {
    return _mapPopover["default"];
  }
});
Object.defineProperty(exports, "MapControlFactory", {
  enumerable: true,
  get: function get() {
    return _mapControl["default"];
  }
});
Object.defineProperty(exports, "LayerHoverInfoFactory", {
  enumerable: true,
  get: function get() {
    return _layerHoverInfo["default"];
  }
});
Object.defineProperty(exports, "CoordinateInfoFactory", {
  enumerable: true,
  get: function get() {
    return _coordinateInfo["default"];
  }
});
Object.defineProperty(exports, "ModalDialogFactory", {
  enumerable: true,
  get: function get() {
    return _modalDialog["default"];
  }
});
Object.defineProperty(exports, "DeleteDatasetModalFactory", {
  enumerable: true,
  get: function get() {
    return _deleteDataModal["default"];
  }
});
Object.defineProperty(exports, "DataTableModalFactory", {
  enumerable: true,
  get: function get() {
    return _dataTableModal["default"];
  }
});
Object.defineProperty(exports, "LoadDataModalFactory", {
  enumerable: true,
  get: function get() {
    return _loadDataModal["default"];
  }
});
Object.defineProperty(exports, "ExportImageModalFactory", {
  enumerable: true,
  get: function get() {
    return _exportImageModal["default"];
  }
});
Object.defineProperty(exports, "ExportDataModalFactory", {
  enumerable: true,
  get: function get() {
    return _exportDataModal["default"];
  }
});
Object.defineProperty(exports, "AddMapStyleModalFactory", {
  enumerable: true,
  get: function get() {
    return _addMapStyleModal["default"];
  }
});
Object.defineProperty(exports, "ExportMapModalFactory", {
  enumerable: true,
  get: function get() {
    return _exportMapModal["default"];
  }
});
Object.defineProperty(exports, "ModalTabsFactory", {
  enumerable: true,
  get: function get() {
    return _modalTabs["default"];
  }
});
Object.defineProperty(exports, "LoadStorageMapFactory", {
  enumerable: true,
  get: function get() {
    return _loadStorageMap["default"];
  }
});
Object.defineProperty(exports, "AnimationControlFactory", {
  enumerable: true,
  get: function get() {
    return _animationControl["default"];
  }
});
Object.defineProperty(exports, "SpeedControlFactory", {
  enumerable: true,
  get: function get() {
    return _speedControl["default"];
  }
});
Object.defineProperty(exports, "AnimationPlaybacksFactory", {
  enumerable: true,
  get: function get() {
    return _playbackControls["default"];
  }
});
Object.defineProperty(exports, "FloatingTimeDisplayFactory", {
  enumerable: true,
  get: function get() {
    return _floatingTimeDisplay["default"];
  }
});
Object.defineProperty(exports, "DatasetSquare", {
  enumerable: true,
  get: function get() {
    return _styledComponents["default"];
  }
});
Object.defineProperty(exports, "TimeWidgetFactory", {
  enumerable: true,
  get: function get() {
    return _timeWidget["default"];
  }
});
Object.defineProperty(exports, "SingleSelectFilterFactory", {
  enumerable: true,
  get: function get() {
    return _singleSelectFilter["default"];
  }
});
Object.defineProperty(exports, "MultiSelectFilterFactory", {
  enumerable: true,
  get: function get() {
    return _multiSelectFilter["default"];
  }
});
Object.defineProperty(exports, "TimeRangeFilterFactory", {
  enumerable: true,
  get: function get() {
    return _timeRangeFilter["default"];
  }
});
Object.defineProperty(exports, "RangeFilterFactory", {
  enumerable: true,
  get: function get() {
    return _rangeFilter["default"];
  }
});
Object.defineProperty(exports, "injector", {
  enumerable: true,
  get: function get() {
    return _injector.injector;
  }
});
Object.defineProperty(exports, "withState", {
  enumerable: true,
  get: function get() {
    return _injector.withState;
  }
});
Object.defineProperty(exports, "CloudTile", {
  enumerable: true,
  get: function get() {
    return _cloudTile["default"];
  }
});
Object.defineProperty(exports, "FileUploadFactory", {
  enumerable: true,
  get: function get() {
    return _fileUpload["default"];
  }
});
Object.defineProperty(exports, "FileUpload", {
  enumerable: true,
  get: function get() {
    return _fileUpload.FileUpload;
  }
});
Object.defineProperty(exports, "DatasetLabel", {
  enumerable: true,
  get: function get() {
    return _datasetLabel["default"];
  }
});
Object.defineProperty(exports, "ItemSelector", {
  enumerable: true,
  get: function get() {
    return _itemSelector["default"];
  }
});
Object.defineProperty(exports, "FieldSelector", {
  enumerable: true,
  get: function get() {
    return _fieldSelector["default"];
  }
});
Object.defineProperty(exports, "RangeSlider", {
  enumerable: true,
  get: function get() {
    return _rangeSlider["default"];
  }
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function get() {
    return _modal["default"];
  }
});
Object.defineProperty(exports, "ModalFooter", {
  enumerable: true,
  get: function get() {
    return _modal.ModalFooter;
  }
});
Object.defineProperty(exports, "ModalTitle", {
  enumerable: true,
  get: function get() {
    return _modal.ModalTitle;
  }
});
Object.defineProperty(exports, "AppLogo", {
  enumerable: true,
  get: function get() {
    return _logo["default"];
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function get() {
    return _switch["default"];
  }
});
Object.defineProperty(exports, "LoadingSpinner", {
  enumerable: true,
  get: function get() {
    return _loadingSpinner["default"];
  }
});
Object.defineProperty(exports, "LoadingDialog", {
  enumerable: true,
  get: function get() {
    return _loadingDialog["default"];
  }
});
exports.Icons = exports.TimeRangeSlider = void 0;

var _timeRangeSlider = _interopRequireDefault(require("./common/time-range-slider"));

var _container = _interopRequireWildcard(require("./container"));

var _keplerGl = _interopRequireDefault(require("./kepler-gl"));

var _sidePanel = _interopRequireWildcard(require("./side-panel"));

var _mapContainer = _interopRequireDefault(require("./map-container"));

var _bottomWidget = _interopRequireDefault(require("./bottom-widget"));

var _modalContainer = _interopRequireDefault(require("./modal-container"));

var _plotContainer = _interopRequireDefault(require("./plot-container"));

var _panelHeader = _interopRequireWildcard(require("./side-panel/panel-header"));

var _panelHeaderAction = _interopRequireDefault(require("./side-panel/panel-header-action"));

var _sideBar = _interopRequireWildcard(require("./side-panel/side-bar"));

var _panelToggle = _interopRequireDefault(require("./side-panel/panel-toggle"));

var _layerManager = _interopRequireWildcard(require("./side-panel/layer-manager"));

var _layerPanel = _interopRequireDefault(require("./side-panel/layer-panel/layer-panel"));

var _layerPanelHeader = _interopRequireDefault(require("./side-panel/layer-panel/layer-panel-header"));

var _layerConfigurator = _interopRequireDefault(require("./side-panel/layer-panel/layer-configurator"));

var _sourceDataCatalog = _interopRequireDefault(require("./side-panel/common/source-data-catalog"));

var _sourceDataSelector = _interopRequireDefault(require("./side-panel/common/source-data-selector"));

var _datasetTitle = _interopRequireDefault(require("./side-panel/common/dataset-title"));

var _datasetInfo = _interopRequireDefault(require("./side-panel/common/dataset-info"));

var _datasetTag = _interopRequireDefault(require("./side-panel/common/dataset-tag"));

var _filterManager = _interopRequireDefault(require("./side-panel/filter-manager"));

var _filterPanel = _interopRequireDefault(require("./side-panel/filter-panel/filter-panel"));

var _interactionManager = _interopRequireDefault(require("./side-panel/interaction-manager"));

var _brushConfig = _interopRequireDefault(require("./side-panel/interaction-panel/brush-config"));

var _tooltipConfig = _interopRequireDefault(require("./side-panel/interaction-panel/tooltip-config"));

var _mapManager = _interopRequireDefault(require("./side-panel/map-manager"));

var _mapLayerSelector = _interopRequireDefault(require("./side-panel/map-style-panel/map-layer-selector"));

var _mapStyleSelector = _interopRequireDefault(require("./side-panel/map-style-panel/map-style-selector"));

var _customPanel = _interopRequireDefault(require("./side-panel/custom-panel"));

var _mapPopover = _interopRequireDefault(require("./map/map-popover"));

var _mapControl = _interopRequireDefault(require("./map/map-control"));

var _layerHoverInfo = _interopRequireDefault(require("./map/layer-hover-info"));

var _coordinateInfo = _interopRequireDefault(require("./map/coordinate-info"));

var _modalDialog = _interopRequireDefault(require("./modals/modal-dialog"));

var _deleteDataModal = _interopRequireDefault(require("./modals/delete-data-modal"));

var _dataTableModal = _interopRequireDefault(require("./modals/data-table-modal"));

var _loadDataModal = _interopRequireDefault(require("./modals/load-data-modal"));

var _exportImageModal = _interopRequireDefault(require("./modals/export-image-modal"));

var _exportDataModal = _interopRequireDefault(require("./modals/export-data-modal"));

var _addMapStyleModal = _interopRequireDefault(require("./modals/add-map-style-modal"));

var _exportMapModal = _interopRequireDefault(require("./modals/export-map-modal/export-map-modal"));

var _modalTabs = _interopRequireDefault(require("./modals/modal-tabs"));

var _loadStorageMap = _interopRequireDefault(require("./modals/load-storage-map"));

var _animationControl = _interopRequireDefault(require("./common/animation-control/animation-control"));

var _speedControl = _interopRequireDefault(require("./common/animation-control/speed-control"));

var _playbackControls = _interopRequireDefault(require("./common/animation-control/playback-controls"));

var _floatingTimeDisplay = _interopRequireDefault(require("./common/animation-control/floating-time-display"));

var _styledComponents = _interopRequireWildcard(require("./common/styled-components"));

Object.keys(_styledComponents).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _styledComponents[key];
    }
  });
});

var _timeWidget = _interopRequireDefault(require("./filters/time-widget"));

var _singleSelectFilter = _interopRequireDefault(require("./filters/single-select-filter"));

var _multiSelectFilter = _interopRequireDefault(require("./filters/multi-select-filter"));

var _timeRangeFilter = _interopRequireDefault(require("./filters/time-range-filter"));

var _rangeFilter = _interopRequireDefault(require("./filters/range-filter"));

var _injector = require("./injector");

var _cloudTile = _interopRequireDefault(require("./modals/cloud-tile"));

var _fileUpload = _interopRequireWildcard(require("./common/file-uploader/file-upload"));

var _datasetLabel = _interopRequireDefault(require("./common/dataset-label"));

var _itemSelector = _interopRequireDefault(require("./common/item-selector/item-selector"));

var _fieldSelector = _interopRequireDefault(require("./common/field-selector"));

var _rangeSlider = _interopRequireDefault(require("./common/range-slider"));

var _modal = _interopRequireWildcard(require("./common/modal"));

var _logo = _interopRequireDefault(require("./common/logo"));

var _switch = _interopRequireDefault(require("./common/switch"));

var _loadingSpinner = _interopRequireDefault(require("./common/loading-spinner"));

var _loadingDialog = _interopRequireDefault(require("./modals/loading-dialog"));

var Icons = _interopRequireWildcard(require("./common/icons"));

exports.Icons = Icons;
// Copyright (c) 2020 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
// Components
// factories
// // side panel factories
// // map factories
// // modal factories
// // common factory
// // Filters
// Injector
// Common Components
var TimeRangeSlider = (0, _timeRangeSlider["default"])();
exports.TimeRangeSlider = TimeRangeSlider;
